### PR TITLE
248 inconsistency between edgepopulationiter connections and edgesiter connections

### PR DESCRIPTION
**Pull request by edasubert**

### CODE REVIEW

File: bluepysnap/edges/edges.py [hunk 1]
```diff
@@ -276,21 +242,13 @@ def iter_connections(
...
                 return_edge_count=return_edge_count,
             )
-            source_pop = pop.source.name
-            target_pop = pop.target.name
-            if return_edge_count:
-                yield from self._add_circuit_ids(it, source_pop, target_pop)
-            elif return_edge_ids:
-                yield from self._add_edge_ids(it, source_pop, target_pop, name)
-            else:
-                yield from self._omit_edge_count(it, source_pop, target_pop)
```
Comment (joni-herttuainen) [inline bluepysnap/edges/edges.py:293]: I'd move this logic to the `EdgePopulation.iter_connections`

**Comment by joni-herttuainen** [comment]
pinged @GianlucaFicarelli  for an extra pair of eyes

File: bluepysnap/edges/edges.py [hunk 2]
```diff
@@ -221,39 +220,6 @@ def pair_edges(self, source_node_id, target_node_id, properties=None):
...
-        )
-
-    @staticmethod
-    def _omit_edge_count(its, source, target):
-        """Generator comprehension adding the CircuitIds to the iterator."""
-        return (
-            (CircuitNodeId(source, source_id), CircuitNodeId(target, target_id))
-            for source_id, target_id in its
-        )
-
```
[the remark this answers is not in the document]
Comment (edasubert) [reply bluepysnap/edges/edges.py:256]: it does look cleaner :+1:

**Review by joni-herttuainen — COMMENTED** [review]
LGTM. Let's discuss the versioning on stand-up on Monday. Otherwise, I'm ready to approve.

File: bluepysnap/edges/edge_population.py [hunk 3]
```diff
@@ -518,6 +518,38 @@ def _optimal_direction():
...
+            (
+                CircuitNodeId(self.source.name, source_id),
+                CircuitNodeId(self.target.name, target_id),
+                CircuitEdgeIds.from_dict({self.name: self.pair_edges(source_id, target_id)}),
+            )
+            for source_id, target_id, _ in its
+        )
+
+    def _omit_edge_count(self, its):
+        """Generator comprehension adding the CircuitIds to the iterator."""
```
Comment (GianlucaFicarelli) [inline bluepysnap/edges/edge_population.py:544]: The docstring should be corrected, it's the same as `_add_circuit_ids`

File: CHANGELOG.rst [hunk 4]
```diff
@@ -1,6 +1,14 @@
 Changelog
 =========
 
+Version v2.2.0~dev0
```
[the remark this answers is not in the document]
Comment (edasubert) [reply CHANGELOG.rst]: I moved it to 3.0.0 since that is my preference, but I will change it according to whatever the agreement will be

**Comment by GianlucaFicarelli** [comment]
LGTM, but before merging, the version 2.1.0 should be released, unless all the new changes are going to be released together in 3.0.0 (but in this case, the changelog should be updated).

Side note, I don't remember if it was discussed if these changes can affect the performance of `iter_connections()`, and at what degree.

**Comment by edasubert** [comment]
Using the same circuit (`"/gpfs/bbp.cscs.ch/project/proj30/home/subert/tickets/NSETM-2250-test-psp/sonata/circuit_sonata.json"`)

on current master:
```
In [3]: len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
Out[3]: 104521

In [4]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.93 s ± 4.07 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.92 s ± 8.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.85 s ± 57.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN"))[:5]
Out[7]: [(14123, 5018), (14123, 5025), (14123, 5026), (14123, 5031), (14123, 5048)]
```

on this branch:
```
In [3]: len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
Out[3]: 104521

In [4]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.9 s ± 1.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.89 s ± 1.44 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit len(list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN")))
1.89 s ± 1.01 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [7]: list(circuit.edges["thalamus_neurons__thalamus_neurons__chemical"].iter_connections("bAC_IN"))[:5]
Out[7]: 
[(CircuitNodeId(population='thalamus_neurons', id=14123),
  CircuitNodeId(population='thalamus_neurons', id=5018)),
 (CircuitNodeId(population='thalamus_neurons', id=14123),
  CircuitNodeId(population='thalamus_neurons', id=5025)),
 (CircuitNodeId(population='thalamus_neurons', id=14123),
  CircuitNodeId(population='thalamus_neurons', id=5026)),
 (CircuitNodeId(population='thalamus_neurons', id=14123),
  CircuitNodeId(population='thalamus_neurons', id=5031)),
 (CircuitNodeId(population='thalamus_neurons', id=14123),
  CircuitNodeId(population='thalamus_neurons', id=5048))]
```

**Comment by edasubert** [comment]
this MR is now blocking my psp-validation upgrade, which is my fault, because I forgot about it, but still it would help me if we could merge this :slightly_smiling_face:

**Comment by joni-herttuainen** [comment]
I'm fine going all the way to 3.0.0 without releasing 2.1.0. AAMOF, since we are going to break stuff, I might just change that deprecated `validate` to completely removed `validate` -> `validate-simulation` + `validate-circuit`.

That being said, I have no issues with the PR.

**Review by joni-herttuainen — APPROVED** [review]

**Pull request merged**